athena-storage-api/src/main/java/com/amazonaws/athena/storage/datasource/CsvDatasource.java

```diff
@@ -0,0 +1,342 @@
+import com.amazonaws.athena.storage.datasource.exception.ReadRecordsException;
+import com.amazonaws.athena.storage.datasource.exception.TableNotFoundException;
+import com.amazonaws.athena.storage.datasource.exception.UncheckedStorageDatasourceException;
+import com.amazonaws.athena.storage.gcs.GcsCsvSplitUtil;
+import com.amazonaws.athena.storage.gcs.GcsInputStream;
+import com.amazonaws.athena.storage.gcs.GroupSplit;
+import com.amazonaws.athena.storage.gcs.StorageSplit;
+import com.amazonaws.athena.storage.gcs.io.FileCacheFactory;
+import com.amazonaws.athena.storage.gcs.io.GcsOfflineStream;
+import com.amazonaws.athena.storage.gcs.io.GcsOnlineStream;
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.Storage;
+import com.google.common.collect.ImmutableList;
+import com.univocity.parsers.common.record.RecordMetaData;
+import com.univocity.parsers.csv.CsvParser;
+import com.univocity.parsers.csv.CsvParserSettings;
+import org.apache.arrow.vector.types.pojo.ArrowType;
+import org.apache.arrow.vector.types.pojo.Field;
+import org.apache.arrow.vector.types.pojo.FieldType;
+import org.apache.arrow.vector.types.pojo.Schema;
+import org.codehaus.jackson.map.ObjectMapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.nio.charset.StandardCharsets;
+import java.text.CharacterIterator;
+import java.text.StringCharacterIterator;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+import static com.amazonaws.athena.storage.StorageConstants.MAX_CSV_FILES_SIZE;
+import static com.amazonaws.athena.storage.StorageConstants.STORAGE_SPLIT_JSON;
+import static com.amazonaws.athena.storage.StorageUtil.getValidEntityName;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class CsvDatasource
+        extends AbstractStorageDatasource
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(CsvDatasource.class);
+
+    // Used by reflection
+    @SuppressWarnings("unused")
+    public CsvDatasource(String gcsCredentialJsonString,
+                         Map<String, String> properties) throws IOException
+    {
+        this(new GcsDatasourceConfig()
+                .credentialsJson(gcsCredentialJsonString)
+                .properties(properties));
+    }
+
+    public CsvDatasource(GcsDatasourceConfig config) throws IOException
+    {
+        super(config);
+    }
+
+    /**
+     * Indicates whether a ths datasource supports grouping of multiple files to form a single table
+     *
+     * @return This datasource doesn't support reading multiple file to form a single table. So it always returns false
+     */
+    @Override
+    public boolean supportsMultiPartFiles()
+    {
+        return false;
+    }
```

Comment:
Not critical, but we could theoretically support this as long as each file had the proper header, right?

Reply:
We'll refactor it during meta-store-like partitioning implementation

Reply:
Now part of partitioning. We no longer need to check this flag. Removed